chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -0,0 +1,637 @@
+        ConnectTransactionError::PoolOwnerRewardCalculationFailed(block_id, pool_id),
+    )?;
+
+    let total_delegations_reward = (total_reward - pool_owner_reward).ok_or(
+        ConnectTransactionError::PoolOwnerRewardCannotExceedTotalReward(
+            block_id,
+            pool_id,
+            pool_owner_reward,
+            total_reward,
+        ),
+    )?;
+
+    // Distribute reward among delegators.
+    // In some cases this process can yield reward unallocated to delegators. This reward goes to the pool owner.
+    let (delegation_undos, unallocated_reward) = if total_delegations_reward > Amount::ZERO {
+        match accounting_adapter.accounting_delta().get_pool_delegations_shares(pool_id)? {
+            Some(delegation_shares) => {
+                let total_delegations_balance =
+                    delegation_shares.values().try_fold(Amount::ZERO, |acc, v| {
+                        (acc + *v).ok_or(ConnectTransactionError::DelegationsRewardSumFailed(
+                            block_id, pool_id,
+                        ))
+                    })?;
+
+                if total_delegations_balance > Amount::ZERO {
+                    let (delegation_undos, delegations_reward_remainder) =
+                        distribute_delegations_pos_reward(
+                            accounting_adapter,
+                            &delegation_shares,
+                            block_id,
+                            pool_id,
+                            total_delegations_balance,
+                            total_delegations_reward,
+                        )?;
+                    (delegation_undos, delegations_reward_remainder)
+                } else {
+                    // If total balance of all delegations is 0 then give the reward to the pool's owner
+                    (Vec::new(), total_delegations_reward)
+                }
+            }
+            // If no delegations then give the reward to the pool's owner
+            None => (Vec::new(), total_delegations_reward),
+        }
+    } else {
+        // Do nothing if no delegations
+        (Vec::new(), Amount::ZERO)
+    };
+
+    let total_owner_reward = (pool_owner_reward + unallocated_reward)
+        .ok_or(ConnectTransactionError::RewardAdditionError(block_id))?;
+    let increase_pool_balance_undo = accounting_adapter
+        .operations(TransactionSource::Chain(block_id))
+        .increase_pool_pledge_amount(pool_id, total_owner_reward)?;
+
+    let undos = delegation_undos
+        .into_iter()
+        .chain(vec![increase_pool_balance_undo].into_iter())
+        .collect();
+
+    Ok(AccountingBlockRewardUndo::new(undos))
+}
+
+fn calculate_pool_owner_reward(total_reward: Amount, pool_data: &PoolData) -> Option<Amount> {
+    (total_reward - pool_data.cost_per_block())
+        .and_then(|v| v * pool_data.margin_ratio_per_thousand().value().into())
+        .and_then(|v| v / 1000)
+        .and_then(|v| v + pool_data.cost_per_block())
```

Comment:
Please add a check at the beginning of this function that cost_per_block < total reward. I can see this being used as an attack vector. Also some tests for this would be nice.

Reply:
I'd also add a check, always, making sure the output of this function is < total reward.

Reply:
> cost_per_block < total reward

not sure that this check is required, because subtraction is already a checked operation. Or do you want it to be super explicit?

Reply:
I see this has already been discussed and added as a TODO. But why not send the entire reward to the owner if `cost_per_block` is greater than `total_reward`? This seems more reasonable than rejecting such blocks.